Add Next.js caching articles

malimalik opened the pull request:
Hey Kyle, 

Added the Next.js caching article, along with the custom visuals for each mechanism within the article.

I've linked both for your convenience.

Article
https://github.com/malimalik/Web-Dev-Simplified-Official-Blog/blob/3aad60f0758e4c93cf38cc7f0c23c3157eeadb8d/src/pages/2023-12/next-js-app-router-cache/index.mdx

Visuals
https://github.com/malimalik/Web-Dev-Simplified-Official-Blog/tree/3aad60f0758e4c93cf38cc7f0c23c3157eeadb8d/public/articleAssets/2023-12/next-js-app-router-cache

WebDevSimplified commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,388 @@
+
+#### Data Cache
+
+The goal of data cache is to minimize the number of calls made to our origin data source and cache the results across incoming server requests and deployments. Next.js achieves this by overriding the native `fetch` API.
```
Again there is no explanation of the problem. This jumps straight into the solution but there is no explanation of what this cache does or why it is important.

WebDevSimplified commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,388 @@
+In this example, we have multiple users visiting our site looking for travel tips about the city of Paris. Our route fetches data from the server like articles, images and recommendations.
+This data does not change often.
+
+When User 1 triggers a `fetch` request, the response is cached inside the data cache and memoized for the same render cycle. This means that the next 4 users who access the same route within a certain period (revalidation period) get served the cached data, resulting in faster response times.
```
You mention revalidation period and specific numbers like 4 people or 1 minute but again you have never explained what these concepts are. Revalidation should not be mentioned until you explain what it is.

WebDevSimplified commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,388 @@
+
+The diagram below demonstrates a step-by-step process of how data cache works.
+
+![data-cache](/articleAssets/2023-12/next-js-app-router-cache/data-cache.png)
```
Can this image be more visual as well.

WebDevSimplified commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,388 @@
+The diagram below demonstrates a step-by-step process of how data cache works.
+
+![data-cache](/articleAssets/2023-12/next-js-app-router-cache/data-cache.png)
+
```
There is nothing in this section about multiple topics:
1. `unstable_cache` from Next.js
2. `revalidateTag`
3. `export const dynamic`
4. `export const revalidate`

WebDevSimplified commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,388 @@
+
+#### Full Route Cache
+
+The idea behind using a full route cache is to optimize both the client-side and the server-side performance. Full route cache deals with caching the React Server Component Payload and HTML.
```
This intro again goes too in depth into individual details of the cache without ever explaining why or what. It also never explains what React Server Component payload is or why that is important until well after the words are used.

WebDevSimplified replied:
The full route cache is a server cache not a client cache.

WebDevSimplified commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,388 @@
+In the above code, `blogData` returns static content, which is cached by default. I have put a `{cache: "force-cache"}` for illustrative purposes but it can be ommited.
+At build time, the RSC Payload and HTML will be cached in the Full Route Cache. This means that when the user runs the application at run-time to access `/page`, they will be served from the Full Route Cache, resulting in faster page loads.
+
+##### Opting out
```
This never mentions how this works with dynamic functions, or dynamic URL params.

WebDevSimplified commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,388 @@
+
+The diagram below demonstrates the step-by-step process of how full route cache works.
+
+![full-route-cache](/articleAssets/2023-12/next-js-app-router-cache/full-route-cache.png)
```
Again make this a visual diagram.

WebDevSimplified commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,388 @@
+
+![full-route-cache](/articleAssets/2023-12/next-js-app-router-cache/full-route-cache.png)
+
+#### Router Cache
```
Again this section is backwards. There is no explanation of the problem or why the cache is useful. Also, there is no explanation of what RSPC is.

WebDevSimplified commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,388 @@
+The diagram below demonstrates the step-by-step process of how full route cache works.
+
+![full-route-cache](/articleAssets/2023-12/next-js-app-router-cache/full-route-cache.png)
+
```
This section should also mention how this cache only works with your production builds since in development all pages are rendered dynamically, thus, they are never stored in this cache.

WebDevSimplified commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,388 @@
+
+
+The diagram below demonstrates the step-by-step process of how the router cache works.
+
```
Overall I felt this section didn't really explain what this cache was. It also never talked about how this cache is invalidated or why it is useful.

WebDevSimplified commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,388 @@
+
+The diagram below demonstrates the step-by-step process of how the router cache works.
+
+![router-cache](/articleAssets/2023-12/next-js-app-router-cache/router-cache.png)
```
Again make this a visual diagram.

malimalik replied:
Done.

WebDevSimplified commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,388 @@
+
+![router-cache](/articleAssets/2023-12/next-js-app-router-cache/router-cache.png)
+
+#### Cache interactions
```
I don't think this section is needed and instead the important information can be moved up into each individual route section.

WebDevSimplified commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,388 @@
+
+The table below summarizes all four caching mechanisms and their details.
+
+| Cache               | Description                                                                         | Location | Enabled by default | How to opt-out                                                                                 | Revalidation Criteria                                                                              |
```
This chart should be modified slightly.
1. The enabled by default column can be removed.
2. The description should be more focused on what the cache does and why that is important (similar to the purpose column in the Next.js docs). (https://nextjs.org/docs/app/building-your-application/caching#overview)
3. You can also probably remove the opt out section

malimalik commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,388 @@
+
+import Tangent from "@blogComponents/lib/Tangent.astro";
+
+### Introduction
```
Resolved.

malimalik commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,388 @@
+
+### Introduction
+
+Next.js is a React framework that offers an extended set of features, including server-side rendering (SSR), file-based routing, caching, and static site generation, all available out of the box.
```
Resolved.

malimalik commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,388 @@
+3. Full Route Cache
+4. Router Cache
+
+<Tangent>
```
Removed.

malimalik commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,388 @@
+
+#### Request Memoization
+
+Let's begin with understanding what memoization is.
```
> This section does not need an explanation on memoization. It is not important to understand what memoization is to understand this section of the caching.
> 
> This section (and all other caching sections) should really focus on a few things.
> 
> What is the purpose of this cache
> Where is the cache stored
> How long does it last for
> How to invalidate the cache
> How to opt out

Changed.

malimalik commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
````diff
@@ -0,0 +1,388 @@
+
+Let's take a look at an example.
+
+```jsx
````
Added with explanation.

malimalik commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
````diff
@@ -0,0 +1,388 @@
+renderPage();
+```
+
+In the code above, `renderPage()` makes two calls to `fetchUserData()`. The first call to `fetchUserData()` is fetched from the data cache or data store. The return value from the first call is then memoized in-memory.
````
Reworded and restructured.

malimalik commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,388 @@
+In the code above, `renderPage()` makes two calls to `fetchUserData()`. The first call to `fetchUserData()` is fetched from the data cache or data store. The return value from the first call is then memoized in-memory.
+When the second call to `fetchUserData()` is made, the return value is retrieved from in-memory cache since it was memoized upon the first call, reducing redundant network calls. Once the route finishes rendering, the memoized values will be invalidated.
+
+Notice that the parameters of both requests are identical. This is crucial for memoization to work - the `url` and `options` parameter must match across all `fetch` requests.
```
Great point, resolved, thanks.

malimalik commented:
@WebDevSimplified, I made some comments and marked conversations as resolved to keep you in the loop about the status of the revision. I have not pushed any new changes yet and will notify you once I do.

malimalik commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,388 @@
+
+I also provide code snippets, accompanied by diagrams that illustrate the step-by-step process of how each mechanism functions.
+
+### Next.js Caching Mechanisms
```
Done.

malimalik commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,388 @@
+
+The diagram below demonstrates the step-by-step process for how request memoization works.
+
+![request-memo](/articleAssets/2023-12/next-js-app-router-cache/request-memoization.png)
```
Done.

malimalik commented:
Revised article added here: 

2f6a8657905dd1939e91e8ed799c6287906d8ad4

WebDevSimplified commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,586 @@
+
+If you don't understand how Next.js's caching mechanism works it feels like you are constantly fighting Next.js instead of reaping the amazing benefits of Next.js's powerful caching. That is why in this article I am going to break down exactly how every part of Next.js's cache works so you can stop fighting it and finally take advantage of its incredible performance gains.
+
+If you are not already familiar with React Server Components, I wrote an entire article on it which explains them in depth. You can read that [here](https://blog.webdevsimplified.com/2023-11/react-server-components/).
```
This doesn't really make sense here as nothing in this section talks about React Server Components. Just remove this.

WebDevSimplified commented on public/articleAssets/2023-12/next-js-app-router-cache/cache-interactions.png:
The text in this image is too small to read. Can you make the font quite a bit larger and easier to read. Also, the word `Cache` does not need to be included in each `Cache SET`/`Cache MISS`/`Cache HIT` (this goes for all the images).

WebDevSimplified commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,586 @@
+
+## Next.js Caching Mechanisms
+
+The main reason Next.js' cache is so hard to understand is because it is
```
This should say `Next.js's` instead of `Next.js'`

WebDevSimplified commented on src/pages/2023-12/next-js-app-router-cache/index.mdx:
```diff
@@ -0,0 +1,586 @@
+3. Full Route Cache
+4. Router Cache
+
+I will be covering each cache's purpose, where it is stored, how long it lasts for, how to invalidate the cache and how to opt out. This seems like a lot of information at first glance but we will take an example based approach.
```
This is missing a comma before `and`. Also you can remove the last sentence from this as well.

WebDevSimplified commented on public/articleAssets/2023-12/next-js-app-router-cache/request-memo.png:
Can the id for the user be changed to `1` as it is easier to see that the ids are the same in both calls.

WebDevSimplified commented on public/articleAssets/2023-12/next-js-app-router-cache/data-cache.png:
This image is wrong. The second user request would go to the data cache and not the request memoization for the cache hit. The entire request memoization section doesn't even need to be included here.

WebDevSimplified replied:
Also this image talks about blog posts in the text which is not what this page shows.

WebDevSimplified reviewed (changes requested):
I am not a huge fan of the fact the images are just direct copies of the Next.js documentation, but they work well so I am ok with them.

I also didn't leave nearly as many comments this time. I think our way of explaining things is just too different for your writing style to work in place of mine so after these revisions are made I will just re-word/re-write this article to fit within my voice and teaching style.

malimalik commented:
Revised article added:

- Changed the font on the cache interactions visual (this is as large as I could get it while ensuring the image doesn't look awkward). In the past, my clients have added a feature to allow users to view the full size on click.
- Rewrote the entire article to match your voice and teaching style.
